File: java-sdk/src/main/java/io/fyno/core_java/FynoUser.java
package io.fyno.core_java;

import io.fyno.core_java.helpers.Config;
import io.fyno.core_java.utils.FynoConstants;
import io.fyno.core_java.utils.FynoContextCreator;
import io.fyno.core_java.utils.FynoUtils;
import io.fyno.core_java.utils.Logger;

import org.json.JSONArray;
import org.json.JSONObject;

public class FynoUser {
    private static final String TAG = "FynoUser";

    private static void updatePush(String tokenType, String token) {
        if (!getIdentity().isEmpty() && !getWorkspace().isEmpty()) {
                try {
                    String ws = getWorkspace();
                    String profile = getIdentity();
                    String endpoint = new FynoUtils().getEndpoint(
                            "update_channel",
                            ws,
                            FynoConstants.ENV,
                            getIdentity(),
                            null,
                            FynoCore.getString("VERSION")
                    );

                    int notificationStatus = FynoCore.areNotificationPermissionsEnabled() ? 1 : 0;
                    JSONObject requestBody = new JSONObject();
                    requestBody.put("channel", new JSONObject().put("push", new JSONArray()
                            .put(new JSONObject()
                                    .put("token", tokenType + ":" + token)
                                    .put("integration_id", getIntegrationId(tokenType))
                                    .put("status", notificationStatus)
                            )
                    ));

                    RequestHandler.requestPOST(endpoint, requestBody, "PATCH");

                    FynoContextCreator.sqlDataHelper.insertConfigByKey(
                            new Config("fyno_" + tokenType + "_token", token)
                    );
                } catch (Exception e) {
                    Logger.i(TAG, "Exception in set" + tokenType.toUpperCase() + "Token: " + e.getMessage());
                }
        }
    }

    private static void updateChannel(String channel, String token) {
        if (!getIdentity().isEmpty() && !getWorkspace().isEmpty()) {
            try {
                String endpoint = new FynoUtils().getEndpoint(
                        "update_channel",
                        getWorkspace(),
                        FynoConstants.ENV,
                        getIdentity(),
                        null,
                        FynoCore.getString("VERSION")
                );

                JSONObject requestBody = new JSONObject();
                requestBody.put("channel", new JSONObject().put(channel, token));

                RequestHandler.requestPOST(endpoint, requestBody, "PATCH");

                FynoContextCreator.sqlDataHelper.insertConfigByKey(
                        new Config("fyno_" + channel, token)
                );
            } catch (Exception e) {
                Logger.i(TAG, "Exception in set" + channel + ": " + e.getMessage());
            }
        }
    }

    private static String getToken(String tokenType) {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_" + tokenType + "_token").getValue();
    }

    private static String getIntegrationId(String tokenType) {
        if ("fcm_token".equals(tokenType)) {
            return getFcmIntegration();
        } else if ("mi_token".equals(tokenType)) {
            return getMiIntegration();
        } else {
            return "";
        }
    }

    public static void identify(String distinctId) {
        FynoContextCreator.sqlDataHelper.insertConfigByKey(new Config("fyno_distinct_id", distinctId));
    }

    public static String getIdentity() {
        String distinctId = FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_distinct_id").getValue();
        return distinctId;
    }

    public static void setFcmToken(String token) {
        updatePush("fcm_token", token);
    }

    public static String getFcmToken() {
        return getToken("fcm_token");
    }

    public static void setMiToken(String token) {
        updatePush("mi_token", token);
    }

    public static String getMiToken() {
        return getToken("mi_token");
    }

    public static void setApnsToken(String token) {
        updatePush("apns_token", token);
    }

    public static String getApnsToken() {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_apns_token").getValue();
    }

    public static void setEmail(String email) {
        updateChannel("email", email);
    }

    public static String getEmail() {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_email").getValue();
    }

    public static void setMobile(String mobile) {
        updateChannel("sms", mobile);
    }

    public static String getMobile() {
        return getToken("mobile");
    }

    public static void setWhatsapp(String mobile) {
        updateChannel("whatsapp", mobile);
    }

    public static String getWhatsapp() {
        return getToken("wa_mobile");
    }

    public static void setWorkspace(String workspaceId) {
        FynoContextCreator.sqlDataHelper.insertConfigByKey(new Config("fyno_wsid", workspaceId));
    }

    public static String getWorkspace() {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_wsid").getValue();
    }

    public static void setFynoIntegration(String integrationId) {
        FynoContextCreator.sqlDataHelper.insertConfigByKey(new Config("fyno_integration_id", integrationId));
    }

    public static String getFynoIntegration() {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_integration_id").getValue();
    }

    public static void setFcmIntegration(String integrationId) {
        FynoContextCreator.sqlDataHelper.insertConfigByKey(new Config("fyno_fcm_integration_id", integrationId));
    }

    public static String getFcmIntegration() {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_fcm_integration_id").getValue();
    }

    public static void setMiIntegration(String integrationId) {
        FynoContextCreator.sqlDataHelper.insertConfigByKey(new Config("fyno_mi_integration_id", integrationId));
    }

    public static String getMiIntegration() {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_mi_integration_id").getValue();
    }

    public static void setApi(String secret) {
        FynoContextCreator.sqlDataHelper.insertConfigByKey(new Config("fyno_ws_secret", secret));
    }

    public static String getApi() {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_ws_secret").getValue();
    }

    public static void setUserName(String name) {
        FynoContextCreator.sqlDataHelper.insertConfigByKey(new Config("fyno_user_name", name));
    }

    public static String getUserName() {
        return FynoContextCreator.sqlDataHelper.getConfigByKey("fyno_user_name").getValue();
    }
}

File: fyno-kotlin-core/src/main/java/io/fyno/core/FynoUser.kt
package io.fyno.core

import android.util.Log
import io.fyno.core.utils.FynoUtils
import io.fyno.core.helpers.Config
import io.fyno.core.utils.FynoConstants
import io.fyno.core.utils.FynoContextCreator
import io.fyno.core.utils.Logger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.json.JSONArray
import org.json.JSONObject

object FynoUser {
    private const val TAG = "FynoUser"
    private fun updatePush(tokenType: String, token: String) {
        val permissions = if (FynoCore.areNotificationPermissionsEnabled()) 1 else 0
        if (getIdentity().isNotEmpty() && getWorkspace().isNotEmpty()) {
            if((permissions.toString() != FynoContextCreator.sqlDataHelper!!.getConfigByKey(
                "fyno_push_permission"
            ).value) || (getIdentity() != FynoContextCreator.sqlDataHelper!!.getConfigByKey(
                    "fyno_push_distinct_id"
                ).value) || (FynoContextCreator.sqlDataHelper!!.getConfigByKey("fyno_push_permission_first_time").value == "false")
            ){
                runBlocking {
                    CoroutineScope(Dispatchers.IO).launch {
                        try {
                            FynoCore.getString("VERSION")
                            Logger.i(TAG, "updatePush: Updating token for user ${getIdentity()}")
                            val endpoint = FynoUtils().getEndpoint(
                                "update_channel",
                                getWorkspace(),
                                env = "live",
                                profile = getIdentity(),
                                newId = null,
                                version = FynoCore.getString("VERSION")
                            )
                            val requestBody = JSONObject().apply {
                                put("channel", JSONObject().apply {
                                    put("push", JSONArray(listOf(JSONObject().apply {
                                        put("token", "$tokenType:$token")
                                        put("integration_id", getFynoIntegration())
                                        put("status", permissions)
                                    })))
                                })
                            }
                            RequestHandler.requestPOST(endpoint, requestBody, "PATCH")
                            FynoContextCreator.sqlDataHelper?.insertConfigByKey(
                                Config(
                                    key = "fyno_${tokenType}_token",
                                    value = token
                                )
                            )
                            FynoContextCreator.sqlDataHelper?.insertConfigByKey(
                                Config(
                                    key = "fyno_push_permission",
                                    value = permissions.toString()
                                )
                            )
                            FynoContextCreator.sqlDataHelper?.insertConfigByKey(
                                Config(
                                    key = "fyno_push_distinct_id",
                                    value = getIdentity()
                                )
                            )
                            FynoContextCreator.sqlDataHelper?.insertConfigByKey(
                                Config(
                                    key = "fyno_push_permission_first_time",
                                    value = "true"
                                )
                            )
                        } catch (e: Exception) {
                            Logger.d(
                                TAG,
                                "Exception in set${tokenType}Token: ${e.message}"
                            )
                        }
                    }
                }
            }
        }
    }

    private fun updateInapp(token: String, integrationId: String){
        if (getIdentity().isNotEmpty() && getWorkspace().isNotEmpty()) {
            runBlocking {
                CoroutineScope(Dispatchers.IO).launch {
                    try {
                        val endpoint = FynoUtils().getEndpoint(
                            "update_channel",
                            getWorkspace(),
                            env = "live",
                            profile = token,
                            newId = null,
                            version = FynoCore.getString("VERSION")
                        )
                        val requestBody = JSONObject().apply {
                            put("channel", JSONObject().apply {
                                put("inapp", JSONArray(listOf(JSONObject().apply {
                                    put("token", token)
                                    put("integration_id", integrationId)
                                    put("status", 1)
                                })))
                            })
                        }
                        RequestHandler.requestPOST(endpoint, requestBody, "PATCH")
                        FynoContextCreator.sqlDataHelper?.insertConfigByKey(
                            Config(
                                key = "fyno_inapp_token",
                                value = token
                            )
                        )
                    } catch (e: Exception) {
                        Logger.d(
                            TAG,
                            "Exception in set Inapp Token: ${e.message}"
                        )
                    }
                }
            }
        }
    }

    private fun updateChannel(channel: String, token: String) {
        if (getIdentity().isNotEmpty() && getWorkspace().isNotEmpty()) {
            runBlocking {
                CoroutineScope(Dispatchers.IO).launch {
                    try {
                        val endpoint = FynoUtils().getEndpoint(
                            "update_channel",
                            getWorkspace(),
                            env = "live",
                            profile = getIdentity(),
                            newId = null,
                            version = FynoCore.getString("VERSION")
                        )
                        val requestBody = JSONObject().apply {
                            put("channel", JSONObject().apply {
                                put(channel, token)
                            })
                        }
                        RequestHandler.requestPOST(endpoint, requestBody, "PATCH")
                        FynoContextCreator.sqlDataHelper?.insertConfigByKey(
                            Config(
                                key = "fyno_${channel}",
                                value = token
                            )
                        )
                    } catch (e: Exception) {
                        Logger.d(TAG, "Exception in set${channel}: ${e.message}")
                    }
                }
            }
        }
    }

    private fun getToken(tokenType: String): String? {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_${tokenType}_token")?.value
    }

    private fun getIntegrationId(tokenType: String): String {
        return when (tokenType) {
            "fcm_token" -> getFcmIntegration()
            "mi_token" -> getMiIntegration()
            else -> ""
        }
    }

    fun identify(distinctId: String) {
        FynoContextCreator.sqlDataHelper?.insertConfigByKey(Config(key = "fyno_distinct_id", value = distinctId))
    }

    fun getIdentity(): String {
        val distinctId = FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_distinct_id")?.value
            ?: ""
        return distinctId
    }

    fun setFcmToken(token: String) {
        updatePush("fcm_token", token)
    }

    fun getFcmToken(): String? {
        return getToken("fcm_token")
    }

    fun setMiToken(token: String) {
        updatePush("mi_token", token)
    }

    fun getMiToken(): String? {
        return getToken("mi_token")
    }

    fun setApnsToken(token: String) {
        updatePush("apns_token", token)
    }

    fun getApnsToken(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_apns_token")?.value ?: ""
    }

    fun setEmail(email: String) {
        updateChannel("email", email)
    }

    fun getEmail(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_email")?.value ?: ""
    }

    fun setMobile(mobile: String) {
        updateChannel("sms", mobile)
    }

    fun getMobile(): String {
        return getToken("mobile") ?: ""
    }

    fun setWhatsapp(mobile: String) {
        updateChannel("whatsapp", mobile)
    }

    fun getWhatsapp(): String {
        return getToken("wa_mobile") ?: ""
    }

    fun setWorkspace(workspaceId: String) {
        FynoContextCreator.sqlDataHelper?.insertConfigByKey(Config(key = "fyno_wsid", value = workspaceId))
    }

    fun getWorkspace(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_wsid")?.value ?: ""
    }

    fun setFynoIntegration(integrationId: String) {
        FynoContextCreator.sqlDataHelper?.insertConfigByKey(Config(key = "fyno_integration_id", value = integrationId))
    }

    fun getFynoIntegration(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_integration_id")?.value ?: ""
    }

    fun setFcmIntegration(integrationId: String) {
        FynoContextCreator.sqlDataHelper?.insertConfigByKey(Config(key = "fyno_fcm_integration_id", value = integrationId))
    }

    fun getFcmIntegration(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_fcm_integration_id")?.value ?: ""
    }

    fun setMiIntegration(integrationId: String) {
        FynoContextCreator.sqlDataHelper?.insertConfigByKey(Config(key = "fyno_mi_integration_id", value = integrationId))
    }

    fun getMiIntegration(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_mi_integration_id")?.value ?: ""
    }

    fun setApi(secret: String) {
        FynoContextCreator.sqlDataHelper?.insertConfigByKey(Config(key = "fyno_ws_secret", value = secret))
    }

    fun getApi(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_ws_secret")?.value ?: ""
    }

    fun setUserName(name: String) {
        FynoContextCreator.sqlDataHelper?.insertConfigByKey(Config(key = "fyno_user_name", value = name))
    }

    fun getUserName(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_user_name")?.value ?: ""
    }

    fun setJWTToken(jwtToken:String){
        FynoContextCreator.sqlDataHelper?.insertConfigByKey(Config(key = "jwt_token", value = jwtToken))
    }

    fun getJWTToken(): String{
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("jwt_token")?.value ?: ""
    }

    fun setInapp(distinct_id: String, integrationId: String){
        updateInapp(distinct_id, integrationId);
    }

    fun getInapp(): String {
        return FynoContextCreator.sqlDataHelper?.getConfigByKey("fyno_inapp_token")?.value ?: ""
    }
}